c10/cuda/CUDACachingAllocator.h

```diff
@@ -118,39 +118,136 @@ struct SegmentInfo {
   std::vector<BlockInfo> blocks;
 };
 
-C10_CUDA_API void* raw_alloc(size_t nbytes);
-C10_CUDA_API void* raw_alloc_with_stream(size_t nbytes, cudaStream_t stream);
-C10_CUDA_API void raw_delete(void* ptr);
-
-C10_CUDA_API Allocator* get();
-C10_CUDA_API void init(int device_count);
-C10_CUDA_API void setMemoryFraction(double fraction, int device);
-C10_CUDA_API void emptyCache();
-C10_CUDA_API void cacheInfo(
-    int dev_id,
-    size_t* cachedAndFree,
-    size_t* largestBlock);
-C10_CUDA_API void* getBaseAllocation(void* ptr, size_t* size);
-C10_CUDA_API void recordStream(const DataPtr&, CUDAStream stream);
-C10_CUDA_API DeviceStats getDeviceStats(int device);
-C10_CUDA_API void resetAccumulatedStats(int device);
-C10_CUDA_API void resetPeakStats(int device);
-C10_CUDA_API std::vector<SegmentInfo> snapshot();
+// Allocator config options.
+enum struct AllocatorBackend : uint8_t {
+  NATIVE = 0,
+  CUDAMALLOCASYNC = 1,
+};
+
+C10_CUDA_API AllocatorBackend allocatorBackend();
+
+// Size pretty-printer
+std::string format_size(uint64_t size);
+
+
+#define FORALL_ALLOCATOR_INTERFACE(_) \
```

Comment:
It'll be helpful to durably record the design considerations that lead to this particular organization for the code. In particular, you should defend why the virtualized interface was done this way, and not more conventionally as just an object with some virtual methods and two implementations.